Sources/Core/Microsoft.StreamProcessing/Aggregates/SlidingMaxAggregate.cs

```diff
@@ -44,51 +52,90 @@ private MinMaxState<T> Accumulate(MinMaxState<T> state, long timestamp, T input)
                 state.currentValue = input;
             }
             return state;
         }
 
+        private void PushToCollection(MinMaxState<T> state)
+        {
+            Contract.Assert(state.currentTimestamp != InvalidSyncTime);
+            Contract.Assert(state.values.Count == 0 || state.values.Last().timestamp <= state.currentTimestamp); // Ensure new timestamp is higher
+
+            int index = state.values.FindLastIndex(e => this.comparer(e.value, state.currentValue) > 0);
+            if (index == -1)
+                state.values.Clear();
+            else
+                state.values.RemoveRange(index + 1, state.values.Count - index - 1);
+
+            state.values.Add(new ValueAndTimestamp<T> { timestamp = state.currentTimestamp, value = state.currentValue });
+            state.currentTimestamp = InvalidSyncTime;
+        }
+
         public Expression<Func<MinMaxState<T>, long, T, MinMaxState<T>>> Deaccumulate()
             => (state, timestamp, input) => state; // never invoked, hence not implemented
 
         public Expression<Func<MinMaxState<T>, MinMaxState<T>, MinMaxState<T>>> Difference()
             => (leftSet, rightSet) => Difference(leftSet, rightSet);
 
-        private static MinMaxState<T> Difference(MinMaxState<T> leftSet, MinMaxState<T> rightSet)
+        private MinMaxState<T> Difference(MinMaxState<T> leftSet, MinMaxState<T> rightSet)
         {
-            if (leftSet.currentTimestamp != InvalidSyncTime)
+            long maxRightTimestamp;
+
+            if (rightSet.currentTimestamp != InvalidSyncTime)
             {
-                leftSet.currentTimestamp = InvalidSyncTime;
-                leftSet.savedValues.Add(leftSet.currentValue);
+                maxRightTimestamp = rightSet.currentTimestamp;
             }
-            if (rightSet.currentTimestamp != InvalidSyncTime)
+            else
+            {
+                if (rightSet.values.Count == 0)
+                    return leftSet;
+
+                // The right set will never contain values for HoppingPipe, adding below for completeness
+                maxRightTimestamp = rightSet.values.Last().timestamp;
+            }
+
+            if (leftSet.currentTimestamp != InvalidSyncTime &&
+                leftSet.currentTimestamp <= maxRightTimestamp)
             {
-                rightSet.currentTimestamp = InvalidSyncTime;
-                rightSet.savedValues.Add(rightSet.currentValue);
+                leftSet.currentTimestamp = InvalidSyncTime; // Discard all values if rightSet's maxTime covers it.
+                leftSet.values.Clear();
+                return leftSet;
+            }
+            else
+            {
+                int maxRange = leftSet.values.FindIndex(e => e.timestamp > maxRightTimestamp);
+                if (maxRange == -1)
+                    leftSet.values.Clear();
+                else
+                    leftSet.values.RemoveRange(0, maxRange);
```

Comment:
>leftSet.values.RemoveRange(0, maxRange); [](start = 20, length = 40)

Hmm, this will copy potentially a large number of elements. I wonder if we can prevent this?